pkg/controller/accept/accept_controller.go

```diff
@@ -0,0 +1,474 @@
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logger.Infof("Dropping cluster out of the queue: %v", err)
+	c.queue.Forget(key)
+}
+
+type jobFactory func(string) (*v1batch.Job, *kapi.ConfigMap, error)
+
+func (f jobFactory) BuildJob(name string) (*v1batch.Job, *kapi.ConfigMap, error) {
+	return f(name)
+}
+
+type jobOwnerControl struct {
+	controller *Controller
+}
+
+func (c *jobOwnerControl) GetOwnerKey(owner metav1.Object) (string, error) {
+	return controller.KeyFunc(owner)
+}
+
+func (c *jobOwnerControl) GetOwner(namespace string, name string) (metav1.Object, error) {
+	return c.controller.clustersLister.Clusters(namespace).Get(name)
+}
+
+func (c *jobOwnerControl) OnOwnedJobEvent(owner metav1.Object) {
+	cluster, ok := owner.(*clusteroperator.Cluster)
+	if !ok {
+		c.controller.logger.WithFields(log.Fields{"owner": owner.GetName(), "namespace": owner.GetNamespace()}).
+			Errorf("attempt to enqueue owner that is not a cluster")
+		return
+	}
+	c.controller.enqueueCluster(cluster)
+}
+
+type jobSyncStrategy struct {
+	controller *Controller
+}
+
+func (s *jobSyncStrategy) GetOwner(key string) (metav1.Object, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if len(namespace) == 0 || len(name) == 0 {
+		return nil, fmt.Errorf("invalid key %q: either namespace or name is missing", key)
+	}
+	return s.controller.clustersLister.Clusters(namespace).Get(name)
+}
+
+func (s *jobSyncStrategy) DoesOwnerNeedProcessing(owner metav1.Object) bool {
+	cluster, ok := owner.(*clusteroperator.Cluster)
+	if !ok {
+		s.controller.logger.Warn("could not convert owner from JobSync into a cluster: %#v", owner)
+		return false
+	}
+	if cluster.Status.Accepted && cluster.Status.AcceptedJobGeneration == cluster.Generation {
+		s.controller.logger.Debugf("cluster %s/%s is already accepted for its current generation, it will not be processed", cluster.Namespace, cluster.Name)
+		return false
+	}
+
+	machineSets, err := controller.MachineSetsForCluster(cluster, s.controller.machineSetsLister)
+	if err != nil {
+		s.controller.logger.Errorf("error retrieving machine sets for cluster %s/%s: %v", cluster.Namespace, cluster.Name, err)
+		utilruntime.HandleError(fmt.Errorf("cannot retrieve machinesets for cluster %s/%s: %v", cluster.Namespace, cluster.Name, err))
+		return false
+	}
+
+	// Only process if all machinesets have been provisioned
+	totalCount := len(machineSets)
+	provisionedCount := 0
+	installedCount := 0
+	for _, ms := range machineSets {
+		if ms.Status.Provisioned {
+			provisionedCount++
+		}
+		if ms.Spec.NodeType == clusteroperator.NodeTypeMaster && ms.Status.Installed {
+			installedCount++
+		}
+	}
+
+	// With at least one machine set, all machineSets in the cluster provisioned and a
+	// master that is installed, we are ready to accept nodes
+	return totalCount > 0 && provisionedCount == totalCount && installedCount > 0
```

Comment:
We also need a check for `len(machineSets) == len(cluster.Spec.MachineSets)` to ensure that all of the machine sets requested in the cluster spec have actually been created.

But as a think about this, I think we need an even stronger guarantee. We need to also ensure that there has not been a change in the cluster spec that is not going to invalidate the machine sets that currently exist. Let's say that the cluster is up and stable for generation 1 of the cluster object. The user changes the cluster version of the cluster, which will cause all of the machine sets to be re-created. The accept controller sees the change to the cluster, notices that the generation is different than the accepted-job generation and runs the accept job again. The job completes successfully, and the accepted-job generation is set to 2. Afterwards, the cluster controller sees the change to the cluster, deletes the machine set and creates new machine sets. The accept controller will not run any accept jobs for the new machine sets because the accepted-job generation makes the current generation of the cluster--which is 2.

I think that ultimately it would be better to have the accept controller sync the machinesets instead of the cluster. (1) That avoids having to keep track of whether the cluster has been updated. (2) We eventually want to be able to accept multiple machine sets anyway.